scripts/patch_websockets.py

```diff
@@ -0,0 +1,91 @@
+# Copyright 2021 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Patches uWebsockets Socket.h to fix a crash sending large
+payloads with ssl encoding.
+"""
+from absl import app
+from absl import flags
+
+FLAGS = flags.FLAGS
+
+flags.DEFINE_string("file", None, "Sockets.h file to patch.")
+
+# The line of Socket.h that we intend to start overwriting, and the 
+# number of lines to overwite.
+REPLACE_BLOCK_START_LINE = 308
+REPLACE_BLOCK_OVERWRITE_LENGTH = 17
+
+# The code which fixes the crash which will replace the existing
+# error handling implementation with one that retries the writes
+# when the SSL encoding buffer is full.
+REPLACE_BLOCK_CODE_LINES = [
+  "              bool continue_loop = true;\n",
+  "              do {\n",
+  "                sent = SSL_write(ssl, message->data, message->length);\n",
+  "                static int bytes_written = 0;\n",
+  "                bytes_written += sent;\n",
+  "                if (sent == (ssize_t) message->length) {\n",
+  "                  wasTransferred = false;\n",
+  "                  return true;\n",
+  "                }\n",
+  "                else if (sent < 0) {\n",
+  "                  switch (SSL_get_error(ssl, sent)) {\n",
+  "                  case SSL_ERROR_WANT_READ:\n",
+  "                    continue_loop = false;\n",
+  "                    break;\n",
+  "                  case SSL_ERROR_WANT_WRITE:\n",
+  "                    break;\n",
```

Comment:
It looks like you removed the entire changePoll() logic.
https://github.com/uNetworking/uWebSockets/blob/4d94401b9c98346f9afd838556fdc7dce30561eb/src/Socket.h#L318

Do you know why that is causing the crash? 
Also, should we replace the part without SSL?
https://github.com/uNetworking/uWebSockets/blob/4d94401b9c98346f9afd838556fdc7dce30561eb/src/Socket.h#L341

Reply:
The changePoll logic was errant. It attempted to delegate to pointer to an Aysnc object that didn't exist, and therefore this is what was actually causing the crash.

I also don't believe the intention of their implemetnation was correct.  The socket shouldn't be set into a read state when SSL_WANT_WRITE is returned.  I've written several MQTT clients which wrapped SLL and network stakcs  I've never seen anything switching to read when the write buffer was full.

Instead code should either loop back and write again. On embeded devices we would throw in a delay and attempt on the next tick, or call select on the underlying socket descriptor to check when its ready to write to again.  This was only an inference that the SSL wrapper had flushed all of its buffered data. to the native socket, though.   

However, in this code we no longer have the socket descriptor to run a select on, and even if we did, we would need to write more code to do this correctly across the the different host opeating systems.  And since this code is  run by us on desktop platforms we can leverage the fact that the SLL implementation  runs in its own process which the OS will timeslice to-- looping on the write should be fine since it loops for only a few moments.   If we were porting this to other devices with slower CPUs or OSes that have scheduling problems then i would suggest extending their implementation to use select, but again the fix would be a lot larger and we'd be heavily meddling with the code of an abandoned implementation, which seems frought with danger in itself.

Also I don't believe we ever use this code in the non ssl mode, so I didn't patch that.